lib/Sema/DerivedConformanceComparable.cpp

```diff
@@ -0,0 +1,411 @@
+
+  // Next, generate the condition being checked.
+  // lhs == rhs
+  auto cmpFuncExpr = new (C) UnresolvedDeclRefExpr(
+    DeclNameRef(C.Id_EqualsOperator), DeclRefKind::BinaryOperator,
+    DeclNameLoc());
+  auto cmpArgsTuple = TupleExpr::create(C, SourceLoc(),
+                                        { lhsExpr, rhsExpr },
+                                        { }, { }, SourceLoc(),
+                                        /*HasTrailingClosure*/false,
+                                        /*Implicit*/true);
+  auto cmpExpr = new (C) BinaryExpr(cmpFuncExpr, cmpArgsTuple,
+                                    /*Implicit*/true);
+  conditions.emplace_back(cmpExpr);
+
+  // Build and return the complete guard statement.
+  // guard lhs == rhs else { return lhs < rhs }
+  auto body = BraceStmt::create(C, SourceLoc(), statements, SourceLoc());
+  return new (C) GuardStmt(SourceLoc(), C.AllocateCopy(conditions), body);
+}
+
+// how does this code ever even get invoked? you can’t compare uninhabited enums...
+static std::pair<BraceStmt *, bool>
+deriveBodyComparable_enum_uninhabited_lt(AbstractFunctionDecl *ltDecl, void *) {
+  auto parentDC = ltDecl->getDeclContext();
+  ASTContext &C = parentDC->getASTContext();
+
+  auto args = ltDecl->getParameters();
+  auto aParam = args->get(0);
+  auto bParam = args->get(1);
+
+  assert(!cast<EnumDecl>(aParam->getType()->getAnyNominal())->hasCases());
+
+  SmallVector<ASTNode, 1> statements;
+  SmallVector<ASTNode, 0> cases;
+
+  // switch (a, b) { }
+  auto aRef = new (C) DeclRefExpr(aParam, DeclNameLoc(), /*implicit*/ true,
+                                  AccessSemantics::Ordinary,
+                                  aParam->getType());
+  auto bRef = new (C) DeclRefExpr(bParam, DeclNameLoc(), /*implicit*/ true,
+                                  AccessSemantics::Ordinary,
+                                  bParam->getType());
+  TupleTypeElt abTupleElts[2] = { aParam->getType(), bParam->getType() };
+  auto abExpr = TupleExpr::create(C, SourceLoc(), {aRef, bRef}, {}, {},
+                                  SourceLoc(), /*HasTrailingClosure*/ false,
+                                  /*implicit*/ true,
+                                  TupleType::get(abTupleElts, C));
+  auto switchStmt = SwitchStmt::create(LabeledStmtInfo(), SourceLoc(), abExpr,
+                                       SourceLoc(), cases, SourceLoc(), C);
+  statements.push_back(switchStmt);
+
+  auto body = BraceStmt::create(C, SourceLoc(), statements, SourceLoc());
+  return { body, /*isTypeChecked=*/true };
+}
+
+/// Derive the body for a '<' operator for an enum that has no associated
+/// values. This generates code that converts each value to its integer ordinal
+/// and compares them, which produces an optimal single icmp instruction.
+static std::pair<BraceStmt *, bool>
+deriveBodyComparable_enum_noAssociatedValues_lt(AbstractFunctionDecl *ltDecl,
+                                               void *) { 
```

Comment:
What's this extra unnamed parameter for?

Reply:
i don’t think it does anything. probably vestigial,, the `deriveBodyEquatable_enum_uninhabited_eq`, `deriveBodyEquatable_enum_noAssociatedValues_eq`, and `deriveBodyEquatable_enum_hasAssociatedValues_eq` functions have them too

Reply:
ah, it’s required by `setBodySynthesizer` on `AbstractFunctionDecl`